web/src/message_events.js

```diff
@@ -430,6 +434,58 @@ export function update_messages(events) {
                 }
             }
 
+            // If the user is narrowed to the stream/topic to which the
+            // messages are moved, we re-narrow it to the same stream/topic
+            // to make sure that messages not available locally are fetched
+            // from the server and shown as expected.
+            if (messages_not_available_locally.length && current_filter) {
+                const stream_name = stream_changed
+                    ? sub_store.get(new_stream_id).name
+                    : old_stream_name;
+                const topic_name = topic_edited ? new_topic : orig_topic;
+
+                if (
+                    current_filter.has_topic(stream_name, topic_name) ||
+                    (current_filter.has_stream(stream_name) &&
+                        !current_filter.has_operator("topic"))
+                ) {
```

Comment:
We should write down why this is the set of cases when we should consider renarrowing. I think the logic is:
- If we don't have the messages, but should have them after the move, then it must be the case that changing the stream or topic of the messages modified whether they match the current view.
- The only way that is possible is if the current view's filters would reject the old stream/topic pair but accept the new one.
- So not only the `stream:` and `topic:` operators are relevant, but also `streams:`, and `search:`.

I think we can go with this logic for the common case for now, but we should have a TODO comment that ideally we'd do the precise check using some logic like `message_matches_search_term` to ask whether it's possible for the move to have previously not matched the narrow but to not match it. Seems like a pain to write and something we can leave for a follow-up, since just stream and topic basic narrows are the 98% case.

Reply:
I had added some explaination above before the first if-block, but have re-phrased it as described above.

Reply:
One point that I forgot to mention before - since we are re-narrowing the list, new `MessageList` object is created and thus we lose some messages that we had fetched previously in the narrow, and we now have only anchor message and 50 messages each before and after the anchor message.